tasks.py

```diff
@@ -231,6 +233,91 @@ def dev_env(ctx, architecture="amd64", name="kind", cni=None):
+
+    # We need the IPs for each Node in the cluster to place them in the BGP
+    # router configuration file (bgpd.conf).  Each Node will peer with this
+    # router.
+    node_ips = run("kubectl get nodes -o jsonpath='{.items[*].status.addresses"
+            "[?(@.type==\"InternalIP\")].address}'", echo=True)
+    node_ips = node_ips.stdout.strip().split()
+    if len(node_ips) != 3:
+        raise Exit(message='Expected 3 nodes, got %d' % len(node_ips))
+
+    # Create a new directory that will be used as the config volume for frr.
+    try:
+        # sudo because past docker runs will have changed ownership of this dir
+        sudo('rm -rf "%s"' % frr_volume_dir)
+        os.mkdir(frr_volume_dir)
+    except FileExistsError:
+        pass
+    except Exception as e:
+        raise Exit(message='Failed to create frr-volume directory: %s'
+                   % str(e))
+
+    # These two config files are static, so we just copy them straight in.
+    shutil.copyfile("%s/frr/zebra.conf" % dev_env_dir,
+                    "%s/zebra.conf" % frr_volume_dir)
+    shutil.copyfile("%s/frr/daemons" % dev_env_dir,
+                    "%s/daemons" % frr_volume_dir)
+
+    # bgpd.conf is created from a template so that we can include the current
+    # Node IPs.
+    with open("%s/frr/bgpd.conf.tmpl" % dev_env_dir, 'r') as f:
+        bgpd_config = "! THIS FILE IS AUTOGENERATED\n" + f.read()
+    for n in range(0, len(node_ips)):
+        bgpd_config = bgpd_config.replace("NODE%d_IP" % n, node_ips[n])
+    with open("%s/bgpd.conf" % frr_volume_dir, 'w') as f:
+        f.write(bgpd_config)
+
+    # Run a BGP router in a container for all of the speakers to peer with.
+    run('([ "$(docker ps -a | grep frr)" ] && docker rm -f frr) || /bin/true')
+    run("docker run -d --privileged --network kind --rm --name frr --volume %s:/etc/frr "
+            "frrouting/frr:latest" % frr_volume_dir, echo=True)
+
+    peer_address = run('docker inspect -f "{{ '
+            'range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}" frr', echo=True)
+    with open("%s/config.yaml.tmpl" % dev_env_dir, 'r') as f:
+        mlb_config = "# THIS FILE IS AUTOGENERATED\n" + f.read()
+    mlb_config = mlb_config.replace("PEER_ADDRESS", peer_address.stdout.strip())
+    with open("%s/config.yaml" % dev_env_dir, 'w') as f:
+        f.write(mlb_config)
+    # Apply the MetalLB ConfigMap
+    run("kubectl apply -f %s/config.yaml" % dev_env_dir)
+
+
+@task
+def dev_env_cleanup(ctx):
+    """Remove traces of the dev env"""
+    try:
+        run("kind delete cluster")
+    except Exception:
+        # This will fail if there's no cluster.
+        pass
+
+    run('([ "$(docker ps -a | grep frr)" ] && docker rm -f frr) || /bin/true')
+
+    dev_env_dir = os.getcwd() + "/dev-env/bgp"
+    frr_volume_dir = dev_env_dir + "/frr-volume"
+
+    # sudo because past docker runs will have changed ownership of this dir
+    sudo('rm -rf "%s"' % frr_volume_dir)
```

Comment:
@russellb This doesn't work here. It needs to ask for a password and immediately fails. From http://docs.pyinvoke.org/en/stable/api/context.html#invoke.context.Context.sudo I see that:

> If you intend to respond to sudo’s password prompt by hand, just use run("sudo command") instead! The autoresponding features in this method will just get in your way.

I have configure to ask for a password here. Switching it to `run('sudo ...')` work fine here. Does it work there too?

Reply:
Ah, good catch!  yes that works for me too